In [1]:
# Imported necessary libraries
import pandas as pd 
import numpy as np

import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

print('succesfully imported libaries....')

succesfully imported libaries....


## Data Preparation

In [2]:
# Loading the dataset
df = pd.read_csv('zillow_data.csv')
df

RegionID  RegionName                 City State              Metro  \
0         84654       60657              Chicago    IL            Chicago   
1         90668       75070             McKinney    TX  Dallas-Fort Worth   
2         91982       77494                 Katy    TX            Houston   
3         84616       60614              Chicago    IL            Chicago   
4         93144       79936              El Paso    TX            El Paso   
...         ...         ...                  ...   ...                ...   
14718     58333        1338             Ashfield    MA    Greenfield Town   
14719     59107        3293            Woodstock    NH          Claremont   
14720     75672       40404                Berea    KY           Richmond   
14721     93733       81225  Mount Crested Butte    CO                NaN   
14722     95851       89155             Mesquite    NV          Las Vegas   

      CountyName  SizeRank   1996-04   1996-05   1996-06  ...  2017-07  \
0           Cook         1  334200.0  335400.0  336500.0  ...  1005500   
1         Collin         2  235700.0  236900.0  236700.0  ...   308000   
2         Harris         3  210400.0  212200.0  212200.0  ...   321000   
3           Cook         4  498100.0  500900.0  503100.0  ...  1289800   
4        El Paso         5   77300.0   77300.0   77300.0  ...   119100   
...          ...       ...       ...       ...       ...  ...      ...   
14718   Franklin     14719   94600.0   94300.0   94000.0  ...   216800   
14719    Grafton     14720   92700.0   92500.0   92400.0  ...   202100   
14720    Madison     14721   57100.0   57300.0   57500.0  ...   121800   
14721   Gunnison     14722  191100.0  192400.0  193700.0  ...   662800   
14722      Clark     14723  176400.0  176300.0  176100.0  ...   333800   

       2017-08  2017-09  2017-10  2017-11  2017-12  2018-01  2018-02  2018-03  \
0      1007500  1007800  1009600  1013300  1018700  1024400  1030700  1033800   
1       310000   312500   314100   315000   316600   318100   319600   321100   
2       320600   320200   320400   320800   321200   321200   323000   326900   
3      1287700  1287400  1291500  1296600  1299000  1302700  1306400  1308500   
4       119400   120000   120300   120300   120300   120300   120500   121000   
...        ...      ...      ...      ...      ...      ...      ...      ...   
14718   217700   218600   218500   218100   216400   213100   209800   209200   
14719   208400   212200   215200   214300   213100   213700   218300   222700   
14720   122800   124600   126700   128800   130600   131700   132500   133000   
14721   671200   682400   695600   695500   694700   706400   705300   681500   
14722   336400   339700   343800   346800   348900   350400   353000   356000   

       2018-04  
0      1030600  
1       321800  
2       329900  
3      1307000  
4       121500  
...        ...  
14718   209300  
14719   225800  
14720   133400  
14721   664400  
14722   357200  

[14723 rows x 272 columns]

In [3]:
# A function to to analyze the shape, number of columns, and information of the dataset
def analyze_dataset(df):
    """
    This function  outputs information about the shape,
    columns, and information of the dataset using the Pandas library.
    """
    # Output the shape of the dataset
    print("Shape of dataset:", df.shape)
    print('\n-----------------------------------------------------------')

    # Output the column names of the dataset
    print("Column names:", list(df.columns))
    print('\n-----------------------------------------------------------')

    # Output information about the dataset
    print(df.info())
    print('\n-----------------------------------------------------------')

    # output descriptive statistics about the dataset
    print(df.describe())
    print('\n-----------------------------------------------------------')


In [4]:
# calling the function
analyze_dataset(df)

Shape of dataset: (14723, 272)

-----------------------------------------------------------
Column names: ['RegionID', 'RegionName', 'City', 'State', 'Metro', 'CountyName', 'SizeRank', '1996-04', '1996-05', '1996-06', '1996-07', '1996-08', '1996-09', '1996-10', '1996-11', '1996-12', '1997-01', '1997-02', '1997-03', '1997-04', '1997-05', '1997-06', '1997-07', '1997-08', '1997-09', '1997-10', '1997-11', '1997-12', '1998-01', '1998-02', '1998-03', '1998-04', '1998-05', '1998-06', '1998-07', '1998-08', '1998-09', '1998-10', '1998-11', '1998-12', '1999-01', '1999-02', '1999-03', '1999-04', '1999-05', '1999-06', '1999-07', '1999-08', '1999-09', '1999-10', '1999-11', '1999-12', '2000-01', '2000-02', '2000-03', '2000-04', '2000-05', '2000-06', '2000-07', '2000-08', '2000-09', '2000-10', '2000-11', '2000-12', '2001-01', '2001-02', '2001-03', '2001-04', '2001-05', '2001-06', '2001-07', '2001-08', '2001-09', '2001-10', '2001-11', '2001-12', '2002-01', '2002-02', '2002-03', '2002-04', '2002-05', '

- The dataset has 14723 rows and 272 columns,4 categorical and the rest are numerical

## Data Cleaning

In [5]:
# Checking for duplicates and missing data in the dataset
def cleaning(data):
    "This is a simple function to get missing and duplicated values"
    missing = data.isna().sum().sum()
    duplicated = data.duplicated().sum()
    return (f"There are '{missing}' missing values and '{duplicated}' duplicated values in the dataset")

In [6]:
# Calling the function
cleaning(df)

"There are '157934' missing values and '0' duplicated values in the dataset"

In [7]:
# creating a dataframe to display the column datatypes and unique values
desc = []
for i in df.columns:
    desc.append([
        i,
        df[i].dtypes,
        df[i].nunique(),
    ])

pd.DataFrame(data = desc, columns=['Feature','Dtypes','Sample_Unique'])

Feature  Dtypes  Sample_Unique
0      RegionID   int64          14723
1    RegionName   int64          14723
2          City  object           7554
3         State  object             51
4         Metro  object            701
..          ...     ...            ...
267     2017-12   int64           5248
268     2018-01   int64           5276
269     2018-02   int64           5303
270     2018-03   int64           5332
271     2018-04   int64           5310

[272 rows x 3 columns]

In [9]:
# Defined a function to  calculate the percentage of missing values in the dataset
def missing_values_percentage(df):
    # Calculating the total number of missing values in the Dataframe
    total_missing = df.isnull().sum().sum()

    # checking the total number of cells in the DataFrame
    total_cells = df.size

    # calculating the percentage of missing values
    percentage_missing = (total_missing / total_cells) * 100
    
    return percentage_missing

missing_values_percentage(df)

3.943759463983923

- Missing values are 3.94% of the entire dataset. Lets go ahead and preview the percentage of missing values per each individual column

In [11]:
# Calculating the percentage of missing values for each column in the DataFrame
missing_values = df.isnull().mean() * 100

# Print the list of columns in the DataFrame along with their missing percentages
for column in missing_values.index:
    # Print the column name followed by its corresponding missing percentage
    print(column, missing_values[column])


RegionID 0.0
RegionName 0.0
City 0.0
State 0.0
Metro 7.084154044691979
CountyName 0.0
SizeRank 0.0
1996-04 7.056985668681655
1996-05 7.056985668681655
1996-06 7.056985668681655
1996-07 7.056985668681655
1996-08 7.056985668681655
1996-09 7.056985668681655
1996-10 7.056985668681655
1996-11 7.056985668681655
1996-12 7.056985668681655
1997-01 7.056985668681655
1997-02 7.056985668681655
1997-03 7.056985668681655
1997-04 7.056985668681655
1997-05 7.056985668681655
1997-06 7.056985668681655
1997-07 7.050193574679074
1997-08 7.050193574679074
1997-09 7.050193574679074
1997-10 7.050193574679074
1997-11 7.050193574679074
1997-12 7.050193574679074
1998-01 7.036609386673912
1998-02 7.036609386673912
1998-03 7.036609386673912
1998-04 7.036609386673912
1998-05 7.036609386673912
1998-06 7.036609386673912
1998-07 7.036609386673912
1998-08 7.036609386673912
1998-09 7.036609386673912
1998-10 7.036609386673912
1998-11 7.036609386673912
1998-12 7.036609386673912
1999-01 7.036609386673912
1999-02 7.0366093

- The percentage of the missing values per column is still low, it is ranging from 1%-7% thus i chose to fill the missing values for the metro column with the string  missing then dropping the missing values in the date columns.

In [15]:
# Filling the 'Metro' column with the word 'missing'
df['Metro'].fillna('missing', inplace=True)

# Calculating the percentage of missing values for each column in the DataFrame
missing_values = df.isnull().mean() * 100

# Handling the date columns' missing values
df.dropna(inplace=True)

# Print the list of columns in the DataFrame along with their missing percentages
for column, percentage in missing_values.items():
    print(f"{column}: {percentage}%")


RegionID: 0.0%
RegionName: 0.0%
City: 0.0%
State: 0.0%
Metro: 0.0%
CountyName: 0.0%
SizeRank: 0.0%
1996-04: 0.0%
1996-05: 0.0%
1996-06: 0.0%
1996-07: 0.0%
1996-08: 0.0%
1996-09: 0.0%
1996-10: 0.0%
1996-11: 0.0%
1996-12: 0.0%
1997-01: 0.0%
1997-02: 0.0%
1997-03: 0.0%
1997-04: 0.0%
1997-05: 0.0%
1997-06: 0.0%
1997-07: 0.0%
1997-08: 0.0%
1997-09: 0.0%
1997-10: 0.0%
1997-11: 0.0%
1997-12: 0.0%
1998-01: 0.0%
1998-02: 0.0%
1998-03: 0.0%
1998-04: 0.0%
1998-05: 0.0%
1998-06: 0.0%
1998-07: 0.0%
1998-08: 0.0%
1998-09: 0.0%
1998-10: 0.0%
1998-11: 0.0%
1998-12: 0.0%
1999-01: 0.0%
1999-02: 0.0%
1999-03: 0.0%
1999-04: 0.0%
1999-05: 0.0%
1999-06: 0.0%
1999-07: 0.0%
1999-08: 0.0%
1999-09: 0.0%
1999-10: 0.0%
1999-11: 0.0%
1999-12: 0.0%
2000-01: 0.0%
2000-02: 0.0%
2000-03: 0.0%
2000-04: 0.0%
2000-05: 0.0%
2000-06: 0.0%
2000-07: 0.0%
2000-08: 0.0%
2000-09: 0.0%
2000-10: 0.0%
2000-11: 0.0%
2000-12: 0.0%
2001-01: 0.0%
2001-02: 0.0%
2001-03: 0.0%
2001-04: 0.0%
2001-05: 0.0%
2001-06: 0.0%
2001-07: 0.0%
2001-

- There are '0' missing values and '0' duplicated values in the dataset.The dataset doesn't have any missing values or any duplicates. Since region ID is the unique identifier, let's check if there is any duplicates in that column.

In [17]:
# Checking for duplicates in the region ID column only
df[df['RegionID'].duplicated(keep=False)]


Empty DataFrame
Columns: [RegionID, RegionName, City, State, Metro, CountyName, SizeRank, 1996-04, 1996-05, 1996-06, 1996-07, 1996-08, 1996-09, 1996-10, 1996-11, 1996-12, 1997-01, 1997-02, 1997-03, 1997-04, 1997-05, 1997-06, 1997-07, 1997-08, 1997-09, 1997-10, 1997-11, 1997-12, 1998-01, 1998-02, 1998-03, 1998-04, 1998-05, 1998-06, 1998-07, 1998-08, 1998-09, 1998-10, 1998-11, 1998-12, 1999-01, 1999-02, 1999-03, 1999-04, 1999-05, 1999-06, 1999-07, 1999-08, 1999-09, 1999-10, 1999-11, 1999-12, 2000-01, 2000-02, 2000-03, 2000-04, 2000-05, 2000-06, 2000-07, 2000-08, 2000-09, 2000-10, 2000-11, 2000-12, 2001-01, 2001-02, 2001-03, 2001-04, 2001-05, 2001-06, 2001-07, 2001-08, 2001-09, 2001-10, 2001-11, 2001-12, 2002-01, 2002-02, 2002-03, 2002-04, 2002-05, 2002-06, 2002-07, 2002-08, 2002-09, 2002-10, 2002-11, 2002-12, 2003-01, 2003-02, 2003-03, 2003-04, 2003-05, 2003-06, 2003-07, 2003-08, 2003-09, 2003-10, 2003-11, 2003-12, ...]
Index: []

[0 rows x 272 columns]

- The data does not have any duplicated ID in the region ID column values

In [18]:
# Defined a function to check and print the value counts for each column in a DataFrame
def check_value_counts(data):
    # Iterate through each column in the DataFrame
    for column in data.columns:
        # Print a message indicating the column name
        print(f'Value counts for {column}:')
        
        # Used the value_counts() function to count unique values in the column and print the result
        print(data[column].value_counts())
        
        # Print a separator line for clarity
        print('------------------------------------------','\n')

# Calling the function
check_value_counts(df)


Value counts for RegionID:
73724    1
72582    1
76672    1
97146    1
91001    1
        ..
62023    1
58806    1
97717    1
60701    1
98304    1
Name: RegionID, Length: 13684, dtype: int64
------------------------------------------ 

Value counts for RegionName:
55324    1
99109    1
47906    1
11040    1
44273    1
        ..
98057    1
89521    1
75182    1
1450     1
65536    1
Name: RegionName, Length: 13684, dtype: int64
------------------------------------------ 

Value counts for City:
New York              104
Los Angeles            93
Houston                86
San Antonio            48
Washington             43
                     ... 
Hastings on Hudson      1
Cold Spring Harbor      1
Saint John              1
Lake Mary               1
Bodega Bay              1
Name: City, Length: 7046, dtype: int64
------------------------------------------ 

Value counts for State:
CA    1188
NY     985
TX     944
PA     797
FL     765
OH     529
IL     523
NJ     502
MI     453
MA    

- The dataset  does not have any data inconsistencies